### PR TITLE
Particle Provider API

### LINKED ISSUE #93: [Request] Particle Factory Registration API
An API to register particle factories on the client. Necessary for any mod that wants to even approach custom particles.

**Forge:** [`ParticleFactoryRegisterEvent`](https://github.com/MinecraftForge/MinecraftForge/blob/1.16.x/src/main/java/net/minecraftforge/client/event/ParticleFactoryRegisterEvent.java) (Forge makes `ParticleManager.registerFactory` accessible)
**Fabric:** [`ParticleFactoryRegistry`](https://github.com/FabricMC/fabric/blob/1.16/fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/client/particle/v1/ParticleFactoryRegistry.java)

~~Additionally, a utility class similar to [`FabricParticleTypes`](https://github.com/FabricMC/fabric/blob/1.16/fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particle/v1/FabricParticleTypes.java) would be appreciated, just so we don't need to extend `DefaultParticleType`.~~ nvm, seems like it's pretty much useless

### PR DESCRIPTION
**Pull request by Leo40Git**
Closes #93.

This API mirrors Fabric's equivalent API, meaning you can easily create animated particles that work on both sides!

### CODE REVIEW

**Comment by Leo40Git** [comment]
I've added a test to the testmod (summon the test particle via `/particle` - I tried adding a block that randomly creates that particle but that failed).

Also, tested on Fabric 1.16.5.

File: common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        throw new AssertionError();
+    }
+
+    @ExpectPlatform
+    public static <T extends ParticleOptions> void register(ParticleType<T> type, PendingParticleProvider<T> constructor) {
+        throw new AssertionError();
+    }
+
+    @FunctionalInterface
+    public interface PendingParticleProvider<T extends ParticleOptions> {
```
Comment (shedaniel) [inline common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java]: DeferredParticleProvider

File: common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        List<TextureAtlasSprite> getSprites();
+    }
+
+    @ExpectPlatform
+    public static <T extends ParticleOptions> void register(ParticleType<T> type, ParticleProvider<T> provider) {
+        throw new AssertionError();
+    }
+
+    @ExpectPlatform
+    public static <T extends ParticleOptions> void register(ParticleType<T> type, PendingParticleProvider<T> constructor) {
```
Comment (shedaniel) [inline common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java]: Param: constructor -> provider
(fix this for the subprojects as well)

**Review by shedaniel — CHANGES_REQUESTED** [review]

**Review by MaxNeedsSnacks — APPROVED** [review]

**Comment by MaxNeedsSnacks** [comment]
Adding last call since Dan‘s review has been resolved

**Review by shedaniel — APPROVED** [review]

**Pull request merged**